### PR TITLE
Add preventCollision option

### PR DESCRIPTION
**Pull request by nikolas**
This option prevents grid items from being moved when collided with. This
is useful functionality for the video sequencer I'm building with
react-grid-layout.

### CODE REVIEW

**Comment by nikolas** [comment]
@STRML this feature has gotten a few upvotes, so it might be a useful addition to RGL for more people than just myself! If there's anything I can do to make it easier for you to provide feedback for this PR, just let me know. If you don't see this as fitting in with the library that's fine too.. I can keep using my branch, but it would be nice to be working off of the upstream version.

**Comment by msiebert** [comment]
@nikolas I tried out this branch, and I think it's really promising :sparkles:. However, I ran into two problems that perhaps you might be able to help me solve.

1) Resizing still nudges other grid items out of the way. I'd love it if the option prevented me from resizing if I were resizing into another grid item.
2) If I move an item over another item that is close to one of the vertical edges of the grid (x=0, for example), and then move it mostly towards the side of that item that borders the edge, the item I'm moving is placed outside the grid.

Any suggestions on how to fix these issues? Thanks much!

**Comment by Jero786** [comment]
@nikolas So, guys, this fix is not ready yet to master branch?

**Comment by Jero786** [comment]
@nikolas @msiebert @STRML 
I test this solution and seems like work but not in all cases. Just an example:
#### Work fine example:
We have for instance to boxes, named `Box1` and `Box2`. When are dragging `Box1` and hit `Box2` into lateral side, work fine (not move `Box2`). 
#### Still not working example:
But if we make to hit `Box1` to `Box2` from top side to down (like hit the `Box1` to top side of `Box2` it still are moving in down direction). 

Guys, any idea?

File: lib/utils.js [hunk 1]
```diff
@@ -241,6 +242,10 @@ export function moveElement(layout: Layout, l: LayoutItem, x: ?number, y: ?numbe
   for (let i = 0, len = collisions.length; i < len; i++) {
     const collision = collisions[i];
     // console.log('resolving collision between', l.i, 'at', l.y, 'and', collision.i, 'at', collision.y);
+    if (preventCollision) {
+      l.x = oldX;
```
Comment (Jero786) [inline lib/utils.js]: In order to not allow to push in vertical direction, please add too: 
```
l.y = oldY;
```

**Comment by Jero786** [comment]
@nikolas @STRML 
Seems wok fine but if you collide from lateral side, but is still displacement if you hit from the top side of the box (which are collide) this box still are moving (by the box which are you dragging). To resolve that just need to add an extra `oldY` position. The same way you did which `oldX`.
 
I left some comments in the PR to be more clear.

And finally! in that way work in all cases that I test so far!
Thanks!

Please let me know when this PR are merged to master branch.

**Comment by nikolas** [comment]
yeah, this patch has been working for me - I've been using it in production since last fall. if anyone has improvements to it feel free to add them.

**Comment by Jero786** [comment]
@nikolas @STRML @msiebert, I found another case that still not working yet.... on vertical resizing :S. If we have and box in the same column below and your box origin (which are dragging) move down and collide which another box. This collider box still displacement. Any suggestion?

**Comment by nikolas** [comment]
My suggestion is to add a patch to the code that fixes your issue

**Comment by Jero786** [comment]
@nikolas  past the diff here:


--- a/lib/ReactGridLayout.jsx
+++ b/lib/ReactGridLayout.jsx
@@ -50,6 +50,8 @@ export default class ReactGridLayout extends React.Component {
 ```  
  // A selector for the draggable handler
     draggableHandle: PropTypes.string,
 
+      preventCollision: PropTypes.bool,
+
     // If true, the layout will compact vertically
     verticalCompact: PropTypes.bool,
 ``` 
@@ -243,7 +245,7 @@ export default class ReactGridLayout extends React.Component {
     };
  ```
     // Move the element to the dragged location.
-    layout = moveElement(layout, l, x, y, true /* isUserAction */);
+    layout = moveElement(layout, l, x, y, true , this.props.preventCollision);
 
     this.props.onDrag(layout, oldDragItem, l, placeholder, e, node);
  ```
@@ -268,7 +270,7 @@ export default class ReactGridLayout extends React.Component {
 ```   
  if (!l) return;
 
     // Move the element here
-    layout = moveElement(layout, l, x, y, true /* isUserAction */);
+    layout = moveElement(layout, l, x, y, true , this.props.preventCollision);
 
     this.props.onDragStop(layout, oldDragItem, l, null, e, node);
  ```
--- a/lib/utils.js
+++ b/lib/utils.js
@@ -217,12 +217,15 @@ export function getStatics(layout: Layout): Array<LayoutItem> {
-export function moveElement(layout: Layout, l: LayoutItem, x: ?number, y: ?number, isUserAction: ?boolean): Layout {
+export function moveElement(layout: Layout, l: LayoutItem, x: ?number, y: ?number, isUserAction: 
 ```
?boolean, preventCollision: ?boolean): Layout {
   if (l.static) return layout;

   // Short-circuit if nothing to do.
   if (l.y === y && l.x === x) return layout;
 
+    const oldX = l.x;
+    const oldY = l.y;
+    const oldH = l.h;
   const movingUp = y && l.y > y;
   // This is quite a bit faster than extending the object
   if (typeof x === 'number') l.x = x;
```
@@ -241,7 +244,12 @@ export function moveElement(layout: Layout, l: LayoutItem, x: ?number, y: ?numbe
```
   for (let i = 0, len = collisions.length; i < len; i++) {
     const collision = collisions[i];
     // console.log('resolving collision between', l.i, 'at', l.y, 'and', collision.i, 'at', collision.y);
-
+      if (preventCollision) {
+          l.x = oldX;
+          l.y = oldY;
+          l.h = oldH;
+          return layout;
+      }
     // Short circuit so we can't infinite loop
     if (collision.moved) continue;
```

**Comment by Rhjulskov** [comment]
Is there any chance, this pull request is getting merged? :)

**Comment by nikolas** [comment]
@Rhjulskov I'm not currently working on this, so I would just open a new pull request with these changes, and include @Jero786's suggestions.

**Comment by STRML** [comment]
Merged in 121b19d5e9ba68a80c8438fa9617ec69e8fa15db

**Pull request closed without merging**